.github/workflows/build-docker-image.yml

```
@@ -1,40 +1,63 @@
 name: Build multi-arch Docker Image
 
 on:
+  release:
+    types: [ published ]
   push:
-    branches: [main]
+    branches: [ main ]
   pull_request:
 
 jobs:
-  build:
+  build-linux-and-windows-multiarch:
     runs-on: ubuntu-latest
     steps:
-    - uses: actions/checkout@v3
-    - uses: actions/setup-go@v3
-      with:
-        go-version: '1.18'
-    - run: go version
-
-    - name: Verify project
-      run: make verify
-
-    - name: Set up QEMU
-      uses: docker/setup-qemu-action@v2
-    - name: Set up Docker Buildx
-      id: buildx
-      uses: docker/setup-buildx-action@v2
-
-    - name: Docker info
-      run: docker info
-    - name: Buildx inspect
-      run: docker buildx inspect
-
-    - name: Build image
-      uses: docker/build-push-action@v3
-      with:
-        context: .
-        file: ./Dockerfile
-        platforms: linux/amd64,linux/arm/v7,linux/arm64,linux/ppc64le,linux/s390x
-        push: false
+      - uses: actions/checkout@v3
+
+      - uses: actions/setup-go@v3
+        with:
+          go-version: '1.18'
+      - run: go version
+
+      - name: modVerify
+        run: go mod verify
+
+      - name: modTidy
+        run: go mod tidy
+
+      - name: ensure compilation
+        env:
+          GOOS: linux
+        run: go build
+
+      - name: Test
+        run: go test -v ./...
+
+      - name: Login to Docker Hub
+        # Only if we need to push an image
+        if: ${{ github.event_name == 'release' && github.event.action == 'published' }}
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          password: ${{ secrets.DOCKERHUB_TOKEN }}
+
+      # Setup for buildx
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Docker info
+        run: docker info
+      - name: Buildx inspect
+        run: docker buildx inspect
+
+      # In case linux/s390x fails with tls issues:
+      # see https://github.com/testcontainers/moby-ryuk/pull/26
+      # and https://github.com/testcontainers/moby-ryuk/pull/40
+      - name: Cross Build for Linux and Windows
         # Use a 'temp' tag, that won't be pushed, for non-release builds
-        tags: testcontainers/ryuk:${{ github.event.release.tag_name || 'temp' }}
+        run: |
+          ./build-multiarch.sh testcontainers/ryuk:${{ github.event.release.tag_name || 'temp' }} ${{ (github.event_name == 'release' && github.event.action == 'published' && 'yes') || 'no' }}
```

Here I'd suggest passing the commit hash instead of `temp`, as we would be able to even push snapshots based on commits, being possible to actually test them before a release.
```suggestion
          ./build-multiarch.sh "testcontainers/ryuk:${{ github.event.release.tag_name || '${GITHUB_SHA}' }}" "${{ (github.event_name == 'release' && github.event.action == 'published' && 'yes') || 'no' }}"
```

With this change, I'd suggest removing the conditional push and always push, listening for the main branch and tags.

I wonder if it could be useful to also push PRs, using "pr-ID" as tag 🤔 , for the same reason: testing them.

Thoughts?

---

I'd prefer Git hashes, because that's the handle useful for a checkout and more independent from GitHub.

---

We can create `latest` and `$commitSha` tags for commits on the main branch (considering the mutability of the `latest` tag as something already known by the consumers, alongside caching issues when the local latest is outdated), and both `PR-$ID` and `$commitSha` for PRs.

At the end, the storage cost is depreciable and managed by Docker Hub.

---

Yes, I don't have any objections against `PR-$ID` as long as there is a tag with `$commitSha` :)

Regarding the `latest` tag: that not the one being used in Testcontainers, right? As long as Testcontainers itself uses non-mutable tags, I think `latest` can be kept for convenience.

---

All I'm writing here is my personal opinion or experience, because it's up to the maintainers to decide such questions. So please take my answers as input or feedback, and not as decision. :)

---

Indeed, we are never going to use latest in the testcontainers libraries. We can provide a latest pointer to the latest commit in the case somebody want to leaves in the edge (and eventually contribute to the upstream if a bug is found)